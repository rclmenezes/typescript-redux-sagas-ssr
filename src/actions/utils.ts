export type Error<T> = { [P in keyof T]?: string };

export type FormikSetErrors<T> = (errors: Error<T>) => void;

export interface FormikPayload<D> {
  data: D;
  setErrors: FormikSetErrors<D>;
}

export interface FormikTokenPayload<P> extends FormikPayload<P> {
  token: string;
}
