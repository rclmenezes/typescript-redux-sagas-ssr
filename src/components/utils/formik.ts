import { AxiosError } from "axios";

export type Error<T> = { [P in keyof T]?: string | undefined };

type FormikSetErrors<T> = (errors: T) => void;

export function setFormikErrorsFromError<T>(error: AxiosError, setErrors: FormikSetErrors<T>) {
  if (error.response && error.response.status === 400 && error.response.data) {
    setErrors(error.response.data);
  }
}
