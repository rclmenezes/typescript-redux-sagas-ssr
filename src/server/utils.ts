import { NextFunction, Request, Response } from "express";
// import sequelize from "../db/sequelize";

// export function wrapAsync<T>(fn: (req: Request, res: Response) => Promise<T>) {
//   return (req: Request, res: Response, next: NextFunction) => {
//     sequelize
//       .transaction(t => {
//         return fn(req, res);
//       })
//       .catch(next);
//   };
// }

export function wrapAsync<T>(fn: (req: Request, res: Response, next: NextFunction) => Promise<T>) {
  return (req: Request, res: Response, next: NextFunction) => {
    Promise.resolve(fn(req, res, next)).catch(next);
  };
}
