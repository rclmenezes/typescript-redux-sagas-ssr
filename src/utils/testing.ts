import { Href, UnregisterCallback } from "history";

import { PageRoutingProps } from "./routing";

export function createMockRouterProps<P>(params: P, search: string = "") {
  const location = {
    hash: "",
    key: "",
    pathname: "",
    search,
    state: {},
  };

  const props: PageRoutingProps<P> = {
    history: {
      action: "POP",
      block: t => {
        const temp: UnregisterCallback = () => {};
        return temp;
      },
      createHref: t => {
        const temp: Href = "";
        return temp;
      },
      go: num => {},
      goBack: () => {},
      goForward: () => {},
      length: 2,
      listen: t => {
        const temp: UnregisterCallback = () => {};
        return temp;
      },
      location,
      push: () => {},
      replace: () => {},
    },
    location,
    match: {
      isExact: true,
      params,
      path: "",
      url: "",
    },
    staticContext: {
      status: 200,
    },
  };
  return props;
}
