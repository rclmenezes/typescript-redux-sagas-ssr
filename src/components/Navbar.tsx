import * as csstips from "csstips";
import React from "react";
import { Link } from "react-router-dom";
import { Container, Menu } from "semantic-ui-react";
import { style } from "typestyle";

import { UserState } from "../reducers/user";
import { PageRoutingProps } from "../utils/routing";

export interface NavbarProps {
  hideNavbar: true | undefined;
  hideLogin: true | undefined;
  logoutSubmit: () => void;
  routing: PageRoutingProps<any>;
  user: UserState;
}

class Navbar extends React.Component<NavbarProps> {
  logout = () => {
    this.props.logoutSubmit();
  };

  public render() {
    return (
      <Container className={this.props.hideNavbar && style(csstips.invisible)}>
        <Menu secondary>
          <Menu.Item>
            <Link to="/">Home</Link>
          </Menu.Item>
          <Menu.Menu position="right" className={this.props.hideLogin && style(csstips.invisible)}>
            <Menu.Item>
              {this.props.user ? (
                <a onClick={this.logout}>Log Out</a>
              ) : (
                <Link to="/login">Log In</Link>
              )}
            </Menu.Item>
          </Menu.Menu>
        </Menu>
      </Container>
    );
  }
}

export default Navbar;
