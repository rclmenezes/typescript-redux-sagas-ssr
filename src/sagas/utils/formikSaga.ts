import { AxiosError } from "axios";
import { call } from "redux-saga/effects";
import { Action, AsyncActionCreators } from "typescript-fsa";

import { FormikPayload } from "../../actions/utils";
import { createServiceSaga, POST } from "./serviceSaga";

export { FormikPayload };

export type FormikAction<P> = Action<FormikPayload<P>>;
export type FormikAsyncActionCreator<P, S, E> = AsyncActionCreators<FormikPayload<P>, S, E>;

export function* handleFormErrors<P>(action: FormikAction<P>, error: AxiosError) {
  if (error.response && error.response.status === 400 && error.response.data) {
    yield call(action.payload.setErrors, error.response.data);
  }
}

export function createFormikSubmitSaga<P, S, E>(
  actionCreator: FormikAsyncActionCreator<P, S, E>,
  url: string,
  extractHeadersFromAction: ((action: FormikAction<P>) => {}) | null = null,
) {
  return createServiceSaga({
    actionCreator,
    extractDataFromAction: action => action.payload.data,
    extractHeadersFromAction,
    method: POST,
    onError: handleFormErrors,
    url,
  });
}
