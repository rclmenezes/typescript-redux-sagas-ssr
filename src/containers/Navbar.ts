import { connect } from "react-redux";
import { bindActionCreators, Dispatch } from "redux";
import actionCreators from "../actions";
import Navbar from "../components/Navbar";
import { RootState } from "../reducers";

const mapStateToProps = (state: RootState) => ({
  user: state.user,
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      logoutSubmit: actionCreators.logout.started,
    },
    dispatch,
  );

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Navbar);
