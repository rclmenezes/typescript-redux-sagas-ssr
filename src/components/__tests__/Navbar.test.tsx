import { shallow } from "enzyme";
import React from "react";

import { createMockRouterProps } from "../../utils/testing";
import Navbar, { NavbarProps } from "../Navbar";

describe("Test Navbar", () => {
  const routing = createMockRouterProps({});
  const MOCK_LOGOUT_SUBMIT = () => {};
  const MOCK_USER = {
    email: "asdf",
    id: 1,
  };

  const setup = (propOverrides: Partial<NavbarProps> = {}) => {
    const props = {
      hideLogin: undefined,
      hideNavbar: undefined,
      logoutSubmit: MOCK_LOGOUT_SUBMIT,
      routing,
      user: null,
      ...propOverrides,
    };

    const wrapper = shallow(<Navbar {...props} />);
    return { props, wrapper };
  };

  test("Test default render", () => {
    const { wrapper } = setup();
    expect(wrapper.find('Link[children="Log In"]').exists()).toBe(true);
    expect(wrapper.prop("className")).toBeUndefined();
    expect(
      wrapper
        .find("MenuMenu")
        .first()
        .prop("className"),
    ).toBeUndefined();
    expect(wrapper).toMatchSnapshot();
  });

  test("Navbar renders logout with user", () => {
    const { wrapper } = setup({ user: MOCK_USER });
    expect(wrapper.find('a[children="Log Out"]').exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  test("Navbar is invisible when hideNavbar=true", () => {
    const { wrapper } = setup({ hideNavbar: true });
    expect(wrapper.prop("className")).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  test("Navbar is invisible when hideLogin=true", () => {
    const { wrapper } = setup({ hideLogin: true });
    expect(
      wrapper
        .find("MenuMenu")
        .first()
        .prop("className"),
    ).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });
});
