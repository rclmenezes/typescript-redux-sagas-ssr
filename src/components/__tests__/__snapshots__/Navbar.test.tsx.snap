// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Navbar Navbar is invisible when hideLogin=true 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Navbar
    hideLogin={true}
    logoutSubmit={[Function]}
    routing={
      Object {
        "history": Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 2,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "",
            "pathname": "",
            "search": "",
            "state": Object {},
          },
          "push": [Function],
          "replace": [Function],
        },
        "location": Object {
          "hash": "",
          "key": "",
          "pathname": "",
          "search": "",
          "state": Object {},
        },
        "match": Object {
          "isExact": true,
          "params": Object {},
          "path": "",
          "url": "",
        },
        "staticContext": Object {
          "status": 200,
        },
      }
    }
    user={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Menu
        secondary={true}
      >
        <MenuItem>
          <Link
            replace={false}
            to="/"
          >
            Home
          </Link>
        </MenuItem>
        <MenuMenu
          className="f1h65fgz"
          position="right"
        >
          <MenuItem>
            <Link
              replace={false}
              to="/login"
            >
              Log In
            </Link>
          </MenuItem>
        </MenuMenu>
      </Menu>,
      "className": undefined,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <MenuItem>
            <Link
              replace={false}
              to="/"
            >
              Home
            </Link>
          </MenuItem>,
          <MenuMenu
            className="f1h65fgz"
            position="right"
          >
            <MenuItem>
              <Link
                replace={false}
                to="/login"
              >
                Log In
              </Link>
            </MenuItem>
          </MenuMenu>,
        ],
        "secondary": true,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Link
              replace={false}
              to="/"
            >
              Home
            </Link>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Home",
              "replace": false,
              "to": "/",
            },
            "ref": null,
            "rendered": "Home",
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <MenuItem>
              <Link
                replace={false}
                to="/login"
              >
                Log In
              </Link>
            </MenuItem>,
            "className": "f1h65fgz",
            "position": "right",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Link
                replace={false}
                to="/login"
              >
                Log In
              </Link>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Log In",
                "replace": false,
                "to": "/login",
              },
              "ref": null,
              "rendered": "Log In",
              "type": [Function],
            },
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Menu
          secondary={true}
        >
          <MenuItem>
            <Link
              replace={false}
              to="/"
            >
              Home
            </Link>
          </MenuItem>
          <MenuMenu
            className="f1h65fgz"
            position="right"
          >
            <MenuItem>
              <Link
                replace={false}
                to="/login"
              >
                Log In
              </Link>
            </MenuItem>
          </MenuMenu>
        </Menu>,
        "className": undefined,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <MenuItem>
              <Link
                replace={false}
                to="/"
              >
                Home
              </Link>
            </MenuItem>,
            <MenuMenu
              className="f1h65fgz"
              position="right"
            >
              <MenuItem>
                <Link
                  replace={false}
                  to="/login"
                >
                  Log In
                </Link>
              </MenuItem>
            </MenuMenu>,
          ],
          "secondary": true,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Link
                replace={false}
                to="/"
              >
                Home
              </Link>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Home",
                "replace": false,
                "to": "/",
              },
              "ref": null,
              "rendered": "Home",
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <MenuItem>
                <Link
                  replace={false}
                  to="/login"
                >
                  Log In
                </Link>
              </MenuItem>,
              "className": "f1h65fgz",
              "position": "right",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <Link
                  replace={false}
                  to="/login"
                >
                  Log In
                </Link>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Log In",
                  "replace": false,
                  "to": "/login",
                },
                "ref": null,
                "rendered": "Log In",
                "type": [Function],
              },
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Test Navbar Navbar is invisible when hideNavbar=true 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Navbar
    hideNavbar={true}
    logoutSubmit={[Function]}
    routing={
      Object {
        "history": Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 2,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "",
            "pathname": "",
            "search": "",
            "state": Object {},
          },
          "push": [Function],
          "replace": [Function],
        },
        "location": Object {
          "hash": "",
          "key": "",
          "pathname": "",
          "search": "",
          "state": Object {},
        },
        "match": Object {
          "isExact": true,
          "params": Object {},
          "path": "",
          "url": "",
        },
        "staticContext": Object {
          "status": 200,
        },
      }
    }
    user={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Menu
        secondary={true}
      >
        <MenuItem>
          <Link
            replace={false}
            to="/"
          >
            Home
          </Link>
        </MenuItem>
        <MenuMenu
          position="right"
        >
          <MenuItem>
            <Link
              replace={false}
              to="/login"
            >
              Log In
            </Link>
          </MenuItem>
        </MenuMenu>
      </Menu>,
      "className": "f1h65fgz",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <MenuItem>
            <Link
              replace={false}
              to="/"
            >
              Home
            </Link>
          </MenuItem>,
          <MenuMenu
            position="right"
          >
            <MenuItem>
              <Link
                replace={false}
                to="/login"
              >
                Log In
              </Link>
            </MenuItem>
          </MenuMenu>,
        ],
        "secondary": true,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Link
              replace={false}
              to="/"
            >
              Home
            </Link>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Home",
              "replace": false,
              "to": "/",
            },
            "ref": null,
            "rendered": "Home",
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <MenuItem>
              <Link
                replace={false}
                to="/login"
              >
                Log In
              </Link>
            </MenuItem>,
            "className": undefined,
            "position": "right",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Link
                replace={false}
                to="/login"
              >
                Log In
              </Link>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Log In",
                "replace": false,
                "to": "/login",
              },
              "ref": null,
              "rendered": "Log In",
              "type": [Function],
            },
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Menu
          secondary={true}
        >
          <MenuItem>
            <Link
              replace={false}
              to="/"
            >
              Home
            </Link>
          </MenuItem>
          <MenuMenu
            position="right"
          >
            <MenuItem>
              <Link
                replace={false}
                to="/login"
              >
                Log In
              </Link>
            </MenuItem>
          </MenuMenu>
        </Menu>,
        "className": "f1h65fgz",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <MenuItem>
              <Link
                replace={false}
                to="/"
              >
                Home
              </Link>
            </MenuItem>,
            <MenuMenu
              position="right"
            >
              <MenuItem>
                <Link
                  replace={false}
                  to="/login"
                >
                  Log In
                </Link>
              </MenuItem>
            </MenuMenu>,
          ],
          "secondary": true,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Link
                replace={false}
                to="/"
              >
                Home
              </Link>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Home",
                "replace": false,
                "to": "/",
              },
              "ref": null,
              "rendered": "Home",
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <MenuItem>
                <Link
                  replace={false}
                  to="/login"
                >
                  Log In
                </Link>
              </MenuItem>,
              "className": undefined,
              "position": "right",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <Link
                  replace={false}
                  to="/login"
                >
                  Log In
                </Link>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Log In",
                  "replace": false,
                  "to": "/login",
                },
                "ref": null,
                "rendered": "Log In",
                "type": [Function],
              },
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Test Navbar Navbar renders logout with user 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Navbar
    logoutSubmit={[Function]}
    routing={
      Object {
        "history": Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 2,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "",
            "pathname": "",
            "search": "",
            "state": Object {},
          },
          "push": [Function],
          "replace": [Function],
        },
        "location": Object {
          "hash": "",
          "key": "",
          "pathname": "",
          "search": "",
          "state": Object {},
        },
        "match": Object {
          "isExact": true,
          "params": Object {},
          "path": "",
          "url": "",
        },
        "staticContext": Object {
          "status": 200,
        },
      }
    }
    user={
      Object {
        "email": "asdf",
        "id": 1,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Menu
        secondary={true}
      >
        <MenuItem>
          <Link
            replace={false}
            to="/"
          >
            Home
          </Link>
        </MenuItem>
        <MenuMenu
          position="right"
        >
          <MenuItem>
            <a
              onClick={[Function]}
            >
              Log Out
            </a>
          </MenuItem>
        </MenuMenu>
      </Menu>,
      "className": undefined,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <MenuItem>
            <Link
              replace={false}
              to="/"
            >
              Home
            </Link>
          </MenuItem>,
          <MenuMenu
            position="right"
          >
            <MenuItem>
              <a
                onClick={[Function]}
              >
                Log Out
              </a>
            </MenuItem>
          </MenuMenu>,
        ],
        "secondary": true,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Link
              replace={false}
              to="/"
            >
              Home
            </Link>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Home",
              "replace": false,
              "to": "/",
            },
            "ref": null,
            "rendered": "Home",
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <MenuItem>
              <a
                onClick={[Function]}
              >
                Log Out
              </a>
            </MenuItem>,
            "className": undefined,
            "position": "right",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <a
                onClick={[Function]}
              >
                Log Out
              </a>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Log Out",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Log Out",
              "type": "a",
            },
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Menu
          secondary={true}
        >
          <MenuItem>
            <Link
              replace={false}
              to="/"
            >
              Home
            </Link>
          </MenuItem>
          <MenuMenu
            position="right"
          >
            <MenuItem>
              <a
                onClick={[Function]}
              >
                Log Out
              </a>
            </MenuItem>
          </MenuMenu>
        </Menu>,
        "className": undefined,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <MenuItem>
              <Link
                replace={false}
                to="/"
              >
                Home
              </Link>
            </MenuItem>,
            <MenuMenu
              position="right"
            >
              <MenuItem>
                <a
                  onClick={[Function]}
                >
                  Log Out
                </a>
              </MenuItem>
            </MenuMenu>,
          ],
          "secondary": true,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Link
                replace={false}
                to="/"
              >
                Home
              </Link>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Home",
                "replace": false,
                "to": "/",
              },
              "ref": null,
              "rendered": "Home",
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <MenuItem>
                <a
                  onClick={[Function]}
                >
                  Log Out
                </a>
              </MenuItem>,
              "className": undefined,
              "position": "right",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <a
                  onClick={[Function]}
                >
                  Log Out
                </a>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Log Out",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Log Out",
                "type": "a",
              },
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Test Navbar Test default render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Navbar
    logoutSubmit={[Function]}
    routing={
      Object {
        "history": Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 2,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "",
            "pathname": "",
            "search": "",
            "state": Object {},
          },
          "push": [Function],
          "replace": [Function],
        },
        "location": Object {
          "hash": "",
          "key": "",
          "pathname": "",
          "search": "",
          "state": Object {},
        },
        "match": Object {
          "isExact": true,
          "params": Object {},
          "path": "",
          "url": "",
        },
        "staticContext": Object {
          "status": 200,
        },
      }
    }
    user={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Menu
        secondary={true}
      >
        <MenuItem>
          <Link
            replace={false}
            to="/"
          >
            Home
          </Link>
        </MenuItem>
        <MenuMenu
          position="right"
        >
          <MenuItem>
            <Link
              replace={false}
              to="/login"
            >
              Log In
            </Link>
          </MenuItem>
        </MenuMenu>
      </Menu>,
      "className": undefined,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <MenuItem>
            <Link
              replace={false}
              to="/"
            >
              Home
            </Link>
          </MenuItem>,
          <MenuMenu
            position="right"
          >
            <MenuItem>
              <Link
                replace={false}
                to="/login"
              >
                Log In
              </Link>
            </MenuItem>
          </MenuMenu>,
        ],
        "secondary": true,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Link
              replace={false}
              to="/"
            >
              Home
            </Link>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Home",
              "replace": false,
              "to": "/",
            },
            "ref": null,
            "rendered": "Home",
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <MenuItem>
              <Link
                replace={false}
                to="/login"
              >
                Log In
              </Link>
            </MenuItem>,
            "className": undefined,
            "position": "right",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Link
                replace={false}
                to="/login"
              >
                Log In
              </Link>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Log In",
                "replace": false,
                "to": "/login",
              },
              "ref": null,
              "rendered": "Log In",
              "type": [Function],
            },
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Menu
          secondary={true}
        >
          <MenuItem>
            <Link
              replace={false}
              to="/"
            >
              Home
            </Link>
          </MenuItem>
          <MenuMenu
            position="right"
          >
            <MenuItem>
              <Link
                replace={false}
                to="/login"
              >
                Log In
              </Link>
            </MenuItem>
          </MenuMenu>
        </Menu>,
        "className": undefined,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <MenuItem>
              <Link
                replace={false}
                to="/"
              >
                Home
              </Link>
            </MenuItem>,
            <MenuMenu
              position="right"
            >
              <MenuItem>
                <Link
                  replace={false}
                  to="/login"
                >
                  Log In
                </Link>
              </MenuItem>
            </MenuMenu>,
          ],
          "secondary": true,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Link
                replace={false}
                to="/"
              >
                Home
              </Link>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Home",
                "replace": false,
                "to": "/",
              },
              "ref": null,
              "rendered": "Home",
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <MenuItem>
                <Link
                  replace={false}
                  to="/login"
                >
                  Log In
                </Link>
              </MenuItem>,
              "className": undefined,
              "position": "right",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <Link
                  replace={false}
                  to="/login"
                >
                  Log In
                </Link>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Log In",
                  "replace": false,
                  "to": "/login",
                },
                "ref": null,
                "rendered": "Log In",
                "type": [Function],
              },
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
